makeflow/src/makeflow_module_singularity.c

```diff
@@ -0,0 +1,68 @@
+/*
+ Copyright (C) 2018- The University of Notre Dame
+ This software is distributed under the GNU General Public License.
+ See the file COPYING for details.
+ */
+
+#include "stringtools.h"
+#include "xxmalloc.h"
+#include "debug.h"
+#include "batch_wrapper.h"
+
+#include "dag.h"
+#include "dag_file.h"
+#include "makeflow_gc.h"
+#include "makeflow_log.h"
+#include "makeflow_hook.h"
+
+#include <errno.h>
+#include <string.h>
+#include <stdlib.h>
+
+#define CONTAINER_SINGULARITY_SH "./singularity.wrapper.sh"
+
+char *singularity_image = NULL;
+
+static int create( struct jx *hook_args )
+{
+    if(jx_lookup_string(hook_args, "singularity_container_image")){
+        singularity_image = xxstrdup(jx_lookup_string(hook_args, "singularity_container_image"));	
+	}
+
+	return MAKEFLOW_HOOK_SUCCESS;
+}
+
+static int node_submit(struct dag_node *n, struct batch_task *t){
+	struct batch_wrapper *wrapper = batch_wrapper_create();
+	batch_wrapper_prefix(wrapper, CONTAINER_SINGULARITY_SH);
+
+	/* Assumes a /disk dir in the image. */
+	char *cmd = string_format("singularity exec --home $(pwd) --bind $(pwd):/disk %s %s", singularity_image, t->command);
```

Comment:
If you bind the cwd to /disk inside the image, part of the command should be cd-ing into /disk when the container starts. That's why I generally say `singularity exec --blah --blah img sh -c "<actual command>"`